moveit_ros/trajectory_cache/src/trajectory_cache.hpp

```diff
@@ -0,0 +1,261 @@
+ *     - Pose waypoints
+ */
+class TrajectoryCache
+{
+public:
+  // We explicitly need a Node::SharedPtr because warehouse_ros ONLY supports
+  // it...
+  //
+  // TODO: methylDragon -
+  // Use rclcpp::node_interfaces::NodeInterfaces<> once warehouse_ros does.
+  explicit TrajectoryCache(const rclcpp::Node::SharedPtr& node);
+
+  bool init(const std::string& db_path = ":memory:", uint32_t db_port = 0, double exact_match_precision = 1e-6);
+
+  unsigned count_trajectories(const std::string& move_group_namespace);
+
+  unsigned count_cartesian_trajectories(const std::string& move_group_namespace);
+
+  // ===========================================================================
+  // MOTION PLAN TRAJECTORY CACHING
+  // ===========================================================================
+  // TOP LEVEL OPS
+
+  // Fetches all plans that fit within the requested tolerances for start and
+  // goal conditions, returning them as a vector, sorted by some db column.
+  std::vector<warehouse_ros::MessageWithMetadata<moveit_msgs::msg::RobotTrajectory>::ConstPtr>
+  fetch_all_matching_trajectories(const moveit::planning_interface::MoveGroupInterface& move_group,
+                                  const std::string& move_group_namespace,
+                                  const moveit_msgs::msg::MotionPlanRequest& plan_request, double start_tolerance,
+                                  double goal_tolerance, bool metadata_only = false,
+                                  const std::string& sort_by = "execution_time_s");
+
+  // Fetches the best trajectory that fits within the requested tolerances for start
+  // and goal conditions, by some db column.
+  warehouse_ros::MessageWithMetadata<moveit_msgs::msg::RobotTrajectory>::ConstPtr fetch_best_matching_trajectory(
+      const moveit::planning_interface::MoveGroupInterface& move_group, const std::string& move_group_namespace,
+      const moveit_msgs::msg::MotionPlanRequest& plan_request, double start_tolerance, double goal_tolerance,
+      bool metadata_only = false, const std::string& sort_by = "execution_time_s");
+
+  // Put a trajectory into the database if it is the best matching trajectory seen so far.
+  //
+  // Trajectories are matched based off their start and goal states.
+  // And are considered "better" if they have a smaller planned execution time
+  // than matching trajectories.
+  //
+  // Optionally deletes all worse trajectories by default to prune the cache.
+  bool put_trajectory(const moveit::planning_interface::MoveGroupInterface& move_group,
+                      const std::string& move_group_namespace, const moveit_msgs::msg::MotionPlanRequest& plan_request,
+                      const moveit_msgs::msg::RobotTrajectory& trajectory, double execution_time_s,
+                      double planning_time_s, bool delete_worse_trajectories = true);
+
+  // QUERY CONSTRUCTION
+  bool extract_and_append_trajectory_start_to_query(warehouse_ros::Query& query,
```

Comment:
Would it make sense to have separate extract and append function for the code to be more modular?

Reply:
The actual extraction is mostly accessing and repackaging the members of the request struct, so probably not. If we did split it however, that would mean adding and returning transitionary structs, which I think is unnecessary (and also adds another layer of indirection and stack allocation.)

I was considering calling it `append_xxx_to_xxx`, but that's technically less correct because we are not simply appending the message, rather, appending some of its members, after doing some small transformations.

Reply:
I will also probably shift these extract_and_append methods to private, since users are more expected to use the top level ops of fetch/put.